### PR TITLE
[ENG-562] Contact export performance improvements

### PR DESCRIPTION
**Pull request by cykonetic**

### CODE REVIEW

File: Controller/ExportController.php [hunk 1]
```diff
@@ -11,145 +11,152 @@
...
-
-        if (!$countOnly) {
-            if (!empty($params['limit'])) {
-                $q->setMaxResults($params['limit']);
-                if (!empty($params['start'])) {
-                    $q->setFirstResult($params['start']);
-                }
-            }
-        }
+            ->setParameter('false', false, 'boolean')
```
Comment (heathdutton) [inline Controller/ExportController.php:118]: Should be `\PDO::PARAM_BOOL` not 'boolean'

File: Controller/ExportController.php [hunk 2]
```diff
@@ -11,145 +11,152 @@
...
     {
-        $dateFromConverted = \DateTime::createFromFormat('M d, Y H:i:s', $dateFrom.'00:00:00')->format('Y-m-d H:i:s');
-        $dateToConverted   = \DateTime::createFromFormat('M d, Y H:i:s', $dateTo.'00:00:00')->format('Y-m-d H:i:s');
+        $timezone = new \DateTimeZone('UTC');
+
+        $dateFromConverted = new \DateTime($dateFrom, $timezone);
 
-        return ['dateFrom' => $dateFromConverted, 'dateTo' => $dateToConverted];
+        $dateToConverted   = new \DateTime($dateTo, $timezone);
+        $dateToConverted->modify('+1 day');
```
Comment (heathdutton) [inline Controller/ExportController.php:144]: This isn't accurate. It should be `->modify('midnight');`

File: Controller/ExportController.php [hunk 3]
```diff
@@ -11,145 +11,152 @@
...
+     *
+     * @return array
+     */
     private function convertDateParams($dateFrom, $dateTo)
     {
-        $dateFromConverted = \DateTime::createFromFormat('M d, Y H:i:s', $dateFrom.'00:00:00')->format('Y-m-d H:i:s');
-        $dateToConverted   = \DateTime::createFromFormat('M d, Y H:i:s', $dateTo.'00:00:00')->format('Y-m-d H:i:s');
+        $timezone = new \DateTimeZone('UTC');
+
+        $dateFromConverted = new \DateTime($dateFrom, $timezone);
```
Comment (heathdutton) [inline Controller/ExportController.php:141]: Do not need to convert timezone.

File: Controller/ExportController.php [hunk 4]
```diff
@@ -11,145 +11,152 @@
...
 
-        return ['dateFrom' => $dateFromConverted, 'dateTo' => $dateToConverted];
+        $dateToConverted   = new \DateTime($dateTo, $timezone);
```
Comment (heathdutton) [inline Controller/ExportController.php:143]: Do not need to convert timezone.

File: Controller/ExportController.php [hunk 5]
```diff
@@ -11,145 +11,152 @@
...
+        $dateToConverted->modify('+1 day');
+
+        return [
+            $dateFromConverted->format('Y-m-d H:i:s'),
```
Comment (heathdutton) [inline Controller/ExportController.php:147]: Should just be `$dateFromConverted->toUtcString()`
This is how it's done throughout the Mautic codebase and doesn't require you to manually format date strings or change the timezone on the \DateTime object.

File: Controller/ExportController.php [hunk 6]
```diff
@@ -11,145 +11,152 @@
...
-                        fputcsv($handle, array_values($contact));
+                        $values = [$id];
+                        foreach ($fieldNames as $fieldName) {
+                            $values[] = $lead->getFieldValue($fieldName);
+                        }
+                        fputcsv($handle, $values);
                     }
-                    $start += $params['limit'];
-                    $this->container->get('doctrine.orm.entity_manager')->clear();
-                    gc_enable();
```
Comment (heathdutton) [inline Controller/ExportController.php:83]: Is gc_collection not needed? Do you not exceed memory limits after a few hundred batches if we do not clear the entity manager? Are the leads no longer in memory?

**Review by heathdutton — CHANGES_REQUESTED** [review]

File: Controller/ExportController.php (hunk 1 above)
Comment (cykonetic) [reply Controller/ExportController.php:118]: Thats what I thought. Mautic core uses 'boolean'

File: Controller/ExportController.php (hunk 2 above)
Comment (cykonetic) [reply Controller/ExportController.php:144]: midnight is the default time when creating with just date

File: Controller/ExportController.php (hunk 3 above)
Comment (cykonetic) [reply Controller/ExportController.php:141]: I wasn't sure. I'll remove that

File: Controller/ExportController.php (hunk 6 above)
Comment (cykonetic) [reply Controller/ExportController.php:83]: not needed. garbage is not part of the memory limit associated with this process. By looping over the results and reassinging to the same variable, memory limits are conserved. I used the laravel StreamedResponse as a model and this issue was addressed

File: Controller/ExportController.php (hunk 5 above)
Comment (cykonetic) [reply Controller/ExportController.php:147]: ok

**Review by heathdutton — APPROVED** [review]

**Pull request merged**